SWIPs/swip-23-kademlia.md

```diff
@@ -0,0 +1,238 @@
+
+## Motivation
+
+Enable other implementations of the swarm node.
+
+## Backwards Compatibility
+
+N/A
+
+## Test Cases
+
+N/A
+
+## Implementation
+
+N/A
+
+## Copyright
+
+Copyright and related rights waived via CC0
+
+## Specification
+
+The Kademlia is both a database (DHT) of known and connected peers on
+the network, as well as a component for routing and forwarding content
+on the network.
+
+### Records
+
+A peer record *ALWAYS* contains the `Address Pair` and the
+`Capabilities` of the peer. This data structure is known as the
+`BzzAddress`. It serializes as follows:
+
+    BZZADDRESS = LIST(OVERLAYADDRESS UNDERLAYADDRESS CAPABILITIES)
+
+### Connectivity
+
+#### Address matching
+
+For two addresses to match each other, their byte values must be
+identical.
+
+For certain purposes in Swarm *partial addresses* are used. A partial
+address is an address of less than 32 bytes. The bytes of a partial
+address always represent the most significant bytes of the full address
+that the partial address represents. A partial address matches a full
+address if all the bytes in the partial address are identical to the
+corresponding bytes in the full address.
+
+#### Distance and proximity
+
+In Swarm, differences between addresses are described in terms of
+*Distance* and *Proximity*.
+
+The *Distance* between two addresses is the value of the big-endian
+binary integer cast of the XOR of the two addresses.
+
+For example, the distance between the addresses 1010 and 1101 will be
+\(1011 ^ 0111 = 1100\), where the big-endian binary integer cast of
+\(1100\) is \(12\).
+
+Another measure of difference *Proximity*, which is the logarithmic
+distance between the addresses. Specifically we define the Proximity
+Order, which is index value of the most significant bit that differs
+between the two addresses.
+
+For example, an address \(0101\) will have a Proximity Order of \(0\)
+compared to \(1101\), a proximity Order of \(1\) compared to \(0011\),
+and a Proximity Order of \(3\) compared to \(0110\). 
+
+#### Proximity Bin
+
+A Proximity Bin is the table row in the kademlia that contains all peers
+that have the corresponding proximity to the node.
+[0.2.2](#sec:distance-and-proximity)
+
+As with the Proximity Order, a *closer* Proximity Bin has a higher
+numeric value, and a *farther* Proximity Bin has a lower numeric value.
+
+#### Neighborhood
+
+The Neighborhood of a node are the peers in the two closest Proximity
+Bins in which peers are known to the node.
+
+#### Depth
+
+The Depth of a node’s kademlia is found by the following method:
+
+1.  Find the two closest Proximity Bins that have at least one
+    *connected* peer.
+
+2.  Find the next farther Proximity Bin that has at least one
+    *connected* peer.
+
+3.  If found, the Depth is the numeric value of that bin plus one.
+
+4.  If not found, the depth is \(0\).
+
+#### Health
+
+A kademlia is healthy when:\[1\]
```

Comment:
if this is the same as kademlia connectivity, I'd suggest calling that out directly.